app/src/main/java/com/github/se/travelpouch/model/activity/map/DirectionsViewModel.kt

```diff
@@ -41,6 +43,28 @@ class DirectionsViewModel(private val repository: DirectionsRepositoryInterface)
         }
   }
 
+  /** Function to fetch directions between activities sequentially. */
+  fun fetchDirectionsForActivities(activities: List<Activity>, mode: String, apiKey: String) {
+    if (activities.size < 2) {
+      Log.e("DirectionsViewModel", "Not enough activities to create a route")
+      return
+    }
+
+    // Extract the origin from the first activity and the destination from the last activity
+    val origin = LatLng(activities.first().location.latitude, activities.first().location.longitude)
+    val destination =
+        LatLng(activities.last().location.latitude, activities.last().location.longitude)
+
+    // Extract the waypoints from the activities in between
+    val waypoints =
+        activities.subList(1, activities.size - 1).map {
+          LatLng(it.location.latitude, it.location.longitude)
+        }
```

Comment:
Again it can be relevant to check null, as in the code below:
```suggestion
    val waypoints = activities.subList(1, activities.size - 1)
         .mapNotNull { it.location?.let { loc -> LatLng(loc.latitude, loc.longitude) } }
```

Reply:
Nice catch but Yassine ensured that the location is never null when we create an activity

Reply:
Yes, it's true by construction. It's more secure and strong to verify it. Free to you to keep it like that or not. I agree with both.

Reply:
Yes you are right. It will be better to verify it.